content/docs/reference/design-docs/train-sim-v3/_index.en.md

```diff
@@ -0,0 +1,514 @@
+- margins
+
+
+There are two types of driving instructions:
+
+- **Abstract driving instructions** model the high-level, rolling stock independent
+range of acceptable behavior: reach 30km/h at this location
+- **Concrete driving instructions** model the specific range of acceptable behavior
+for a specific rolling stock, using limit curves: don't go faster than this curve
+
+```mermaid
+flowchart TD
+Constraint[constraint]
+AbstractDrivingInstruction[abstract driving instruction]
+ConcreteDrivingInstruction[concrete driving instruction]
+RollingStockIntegrator[rolling stock integrator]
+Compiler([compiler])
+
+Constraint -- generates one or more --> AbstractDrivingInstruction
+AbstractDrivingInstruction --> Compiler
+RollingStockIntegrator --> Compiler
+Compiler --> ConcreteDrivingInstruction
+```
+
+
+#### Interpreting driving instructions
+
+During the simulation, driving instructions are partitionned into 4 sets:
+
+- `PENDING` instructions may apply at some point in the future
+- `RECEIVED` instructions aren't enforced yet, but will be unless overridden
+- `ENFORCED` instructions influence train behavior
+- `DISABLED` instructions don't ever have to be considered anymore. There are multiple ways instructions can be disabled:
+  - `SKIPPED` instructions were not received
+  - `RETIRED` instructions expired by themselves
+  - `OVERRIDEN` instructions were removed by another instruction
+
+```mermaid
+flowchart TD
+
+subgraph disabled
+    skipped
+    retired
+    overriden
+end
+
+subgraph active
+    received
+    enforced
+end
+
+pending --> received
+pending --> skipped
+received --> enforced
+received --> overriden
+enforced --> retired
+enforced --> overriden
+```
+
+These sets evolve as follows:
+
+- when an integration steps overlaps a `PENDING` instruction's received condition, it is `RECEIVED` and becomes a candidate to execution
+  - existing instructions may be `OVERRIDEN` due to an `override_on_received` operation
+- if an instruction cannot ever be received at any future simulation state, it transitions to the `SKIPPED` state
+- when simulation state exceeds an instruction's enforcement position, it becomes `ENFORCED`. Only enforced instructions influence train behavior.
+  - existing instructions may be `OVERRIDEN` due to an `override_on_enforced` operation
+- when simulation state exceeds an instruction's retirement position, it becomes `RETIRED`
+
+
+#### Overrides
+
+When an instruction transitions to the `RECEIVED` or `ENFORCED` state, it can disable active instructions
+which match some metadata predicate. There are two metadata attributes which can be relied on for overrides:
+
+- the `kind` allows overriding previous instructions for a given domain, such as spacing or block signaled speed limits
+- the `rank` can be used as a "freshness" or "priority" field. If two instructions overriding each other are received
+  (such as when a train sees two signals), the rank allows deciding which instruction should be prioritized.
```

Comment:
Just a remark: I don't remember the ranking system from our meetings, but I think it'd be better to have another system for priority resolving. Using ints rather than something semantically meaningful seems flaky to me. Maybe we'll find something different when implementing.

Reply:
I do remember we explored a number of other options. I can add a section about what I remember of this debate. We didn't really _like_ this solution, but every other was either way worse or did not work.

Reply:
The decision process is documented a bit further down in the doc I believe